lib/qos/qos.c

```diff
@@ -0,0 +1,356 @@
+LOG_MODULE_REGISTER(qos, CONFIG_QOS_LOG_LEVEL);
+
+/* Define a custom STATIC macro that exposes internal variables when unit testing. */
+#if defined(CONFIG_UNITY)
+#define STATIC
+#else
+#define STATIC static
+#endif
+
+/* Structure used to keep track of pending messages. Used in combination with a linked list. */
+struct qos_metadata {
+	/* Mandatory variable used to construct a linked list. */
+	sys_snode_t header;
+
+	/* Message associated with each node in the linked list. */
+	struct qos_data message;
+};
+
+/* Structure containing internal variables in the library.  */
+STATIC struct ctx {
+	/* Library event handler. Used in callbacks to the caller. */
+	qos_evt_handler_t app_evt_handler;
+
+	/* Internal array of pending messages. Used in combination with linked list. */
+	struct qos_metadata list_internal[CONFIG_QOS_PENDING_MESSAGES_MAX];
+
+	/* Linked list used to keep track of pending messages. */
+	sys_slist_t pending_list;
+
+	/* Variable used to prevent multiple library initializations. */
+	bool initialized;
+
+	/* Delayed work used as the backoff timer. */
+	struct k_work_delayable timeout_handler_work;
+
+	/* Internal variable used to generate message IDs. */
+	uint16_t message_id_next;
+} ctx = {
+	.message_id_next = QOS_MESSAGE_ID_BASE
+};
+
+/* Mutex used to protect internal library variables. */
+static K_MUTEX_DEFINE(ctx_lock);
+
+/* Internal API functions */
+
+static void notify_event(struct qos_evt *evt)
+{
+	__ASSERT(evt != NULL, "Library event not found");
+
+	if (ctx.app_evt_handler != NULL) {
+		ctx.app_evt_handler(evt);
+	}
+}
+
+STATIC void timeout_handler_work_fn(struct k_work *work)
+{
+	struct qos_metadata *node = NULL, *next_node = NULL;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST
+	};
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	/* Remove all messages where the notified_count equals or exceeds
+	 * CONFIG_QOS_MESSAGE_NOTIFIED_COUNT_MAX
+	 */
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ctx.pending_list, node, next_node, header) {
+		if (node->message.notified_count >= CONFIG_QOS_MESSAGE_NOTIFIED_COUNT_MAX) {
+			LOG_DBG("Notified count for message ID: %d exceeds the maximum allowed "
+				"value, remove message from pending list.", node->message.id);
+			evt.message = node->message;
+			notify_event(&evt);
+			memset(&node->message, 0, sizeof(struct qos_data));
+			sys_slist_find_and_remove(&ctx.pending_list, &node->header);
+		}
+	};
+
+	/* Don't shedule a new work if the pending list is empty. */
+	if (sys_slist_is_empty(&ctx.pending_list)) {
+		LOG_DBG("QoS list is empty, don't reschedule work");
+		goto exit;
+	}
+
+	/* Notify all remaining entries in the list. */
+	evt.type = QOS_EVT_MESSAGE_TIMER_EXPIRED;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ctx.pending_list, node, next_node, header) {
+		node->message.notified_count++;
+		evt.message = node->message;
+		notify_event(&evt);
+	};
+
+	k_work_reschedule(&ctx.timeout_handler_work,
+			  K_SECONDS(CONFIG_QOS_MESSAGE_NOTIFY_TIMEOUT_SECONDS));
```

Comment:
If the intention is to use this library for future protocols like LWM2M and CoAP (which usually underlies LWM2M), then I think you should be offering an exponential back-off option here.  See: https://datatracker.ietf.org/doc/html/rfc7252#section-4.2.

Reply:
I initially had an exponential backoff, but chose to not include it due to it being a little more tricky to implement. But, in general I agree that it would be beneficial to have. Please see this [discussion](https://github.com/nrfconnect/sdk-nrf/pull/6892#discussion_r832121215). So for now its a future improvement.

The upcoming integration with lwm2m will not use the QoS library since the lwm2m zephyr API has its own internal retransmission routines.